pkg/sign/signer.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sign
+
+import (
+	"crypto"
+	"crypto/rand"
+	_ "crypto/sha256" // hash function will be used by Sign
+	_ "crypto/sha512" // hash function will be used by Sign
+	"errors"
+
+	protobundle "github.com/sigstore/protobuf-specs/gen/pb-go/bundle/v1"
+	protocommon "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+)
+
+type Signer interface {
+	Sign(data []byte) (*protobundle.Bundle, error)
+}
+
+// TODO: implement
+// type Fulcio struct {
+//	baseURL       string
+//	identityToken string
+// }
+
+// func (f *Fulcio) Sign(data []byte) (*protobundle.Bundle, error) {}
+
+type Keypair struct {
+	signer        crypto.Signer
+	hashAlgorithm protocommon.HashAlgorithm
+	publicKeyHint []byte
+}
+
+func SignerKeypair(s crypto.Signer, ha protocommon.HashAlgorithm, pkh []byte) Keypair {
+	return Keypair{signer: s, hashAlgorithm: ha, publicKeyHint: pkh}
```

Comment:
For `publicKeyHint`, can we make this an optional parameter, and if it's not given, derive it ourselves (`SHA-256(DER(SPKI(pub key)))`)? Most clients won't be opinionated on the key hint I suspect.

Reply:
I'm not sure I produced the hint in exactly the way you suggested, but we do now generate a hint string if one is not provided.

Reply:
Looks great! That's equivalent, DER(SPKI()) is the same as a marshaled PKIX key.